src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,177 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.next = next;
+            this.value = value;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node;
+        if (size == 0) {
+            node = new Node<>(null, value, null);
+            head = node;
+        } else {
+            node = new Node<>(tail, value, null);
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexOutOfBoundsExceptionForAdd(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> indexNode;
+        Node<T> newNode;
+        if (index == 0) {
+            newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+            return;
+        }
+        indexNode = findIndexNode(index);
+        newNode = new Node<>(indexNode.prev, value, indexNode);
+        newNode.prev.next = newNode;
+        newNode.next.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBoundsException(index);
+        return findIndexNode(index).value;
```

Comment:
you use them together, would make sense to check index in the retrieval by index (findIndexNode) method